tools/logger/filter.c

```diff
@@ -0,0 +1,333 @@
+ *
+ * Examples:
+ *   `debug="pipe1"` - set debug log level for components from pipeline1
+ *   `d="pipe1, dai2.3"` - as above, but also for dai2.3
+ *   `error="FIR*"` - for each FIR component set log level to error
+ *
+ * @param dictionary uuid dictionary from ldc file
+ * @param input_str log level settings in format `log_level=component`
+ * @param out_list output list with filter_element elements
+ */
+static int filter_parse_entry(const struct snd_sof_uids_header *uids_dict,
+			      char *input_str, struct list_item *out_list)
+{
+	struct filter_element *filter;
+	char *comp_fmt_end;
+	int32_t log_level;
+	char *comp_fmt;
+	int ret;
+
+	/*
+	 * split string on '=' char, left part describes log level,
+	 * the right one is for component description.
+	 */
+	comp_fmt = strchr(input_str, '=');
+	if (!comp_fmt) {
+		log_err(NULL, "unable to find `=` in `%s`\n", input_str);
+		return -EINVAL;
+	}
+	*comp_fmt = 0;
+	++comp_fmt;
+
+	/* find correct log level in given conf string - string before `=` */
+	log_level = filter_parse_log_level(input_str);
+	if (log_level < 0) {
+		log_err(NULL, "unable to parse log level from `%s`\n",
+			input_str);
+		return log_level;
+	}
+
+	/*
+	 * now parse list of components name and optional instance identifier,
+	 * split string on `COMPONENTS_SEPARATOR`
+	 */
+	while (comp_fmt) {
+		filter = malloc(sizeof(struct filter_element));
+		if (!filter) {
+			log_err(NULL, "unable to malloc memory\n");
+			return -ENOMEM;
+		}
+
+		comp_fmt_end = strchr(comp_fmt, COMPONENTS_SEPARATOR);
+		if (comp_fmt_end)
+			*comp_fmt_end = '\0';
+		ret = filter_parse_component(uids_dict, comp_fmt, filter);
+		if (ret < 0) {
+			log_err(NULL, "unable to parse component from `%s`\n",
+				comp_fmt);
+			free(filter);
+			return ret;
+		}
+		filter->log_level = log_level;
+
+		list_item_append(&filter->list, out_list);
+		comp_fmt = comp_fmt_end ? comp_fmt_end + 1 : 0;
+	}
+
+	return 0;
+}
+
+/**
+ * Parse `input_str` content and send it to FW via debugFS.
+ *
+ * `input_str` contain single filter definition element per line.
+ * Each line is parsed by `filter_parse_entry`, and saved in list.
+ * List of `sof_ipc_dma_trace_filter_elem` is writend to debugFS,
+ * and then send as IPC to FW (this action is implemented in driver).
+ * Each line in debugFS represents single IPC message.
+ *
+ * @param dictionary uuid dictionary from ldc file
+ * @param format log level settings in format `log_level=component`
+ */
+int filter_update_firmware(const struct snd_sof_uids_header *uids_dict,
+			   char *input_str)
+{
+	struct filter_element *filter;
+	struct list_item filter_list;
+	struct list_item *list_elem;
+	struct list_item *list_temp;
+	char *line_end;
+	FILE *out_fd;
+	int ret = 0;
+
+	list_init(&filter_list);
+
+	/* parse `input_str` line by line */
+	line_end = strchr(input_str, '\n');
+	while (line_end) {
+		line_end[0] = '\0';
+		ret = filter_parse_entry(uids_dict, input_str, &filter_list);
+		if (ret < 0)
+			goto err;
```

Comment:
As reported by gcc 9.3 -Werr, this leaved out_fd uninitialized and can make fclose() below trying to close a random file descriptor. Fixed by #3412 

By the way are we still stuck with antique start-of-block declarations because of some ancient xtensa toolchain or can we starting using C99's safer variable declarations?

Reply:
@marc-hb the start of block declaration is part of the "safety critical" coding standard. I would suggest speaking to a safety critical expert and see what their take is on C99 declarations. I think last time that @plbossart or I checked this was not yet approved.